Admin View Model for #29 Phase 2

tippyweb opened the pull request:
The following JSPs were not included:
error.jsp
home.jsp
index.jsp
login.jsp
because they are static pages.
@nowucca -- your feedback is appreciated.
Thank you!!

tippyweb committed:
Merge remote-tracking branch 'refs/remotes/nowucca/master'

tippyweb committed:
Admin View Model updates

Admin View Model updates for #29 Phase 2

nowucca commented on src/main/java/viewmodel/admin/AdminOrderViewModel.java:
```diff
@@ -0,0 +1,80 @@
+    public List<CustomerOrderLineItem> getOrderedProducts() {
+        return orderedProducts;
+    }
+
```
getCustomerOrderDetails not required here?

tippyweb replied:
Hello @nowucca 

My apologies!!  I just noticed that there were a few of ${orderedProduct.quantity} without "p." inside order.jsp!!  Not sure why the displays were correct without "p."??

In fact, I also noticed that there was ${initParam.deliverySurcharge/100.0}, which was left out...  But it seems like the delivery surcharge was displayed correctly...  I guess this needs to be corrected as well...  Any advice is appreciated.  Thank you so much again, @nowucca !!!

nowucca replied:
If you want to sweep over it again to fix these thigns, feel free.
I'll test it and merge it on the weekend.
I was not 100% thorough with my review yesterday obviously!
The goal is in fact to get rid of most ${initParam} expressions and use ${p} for most things, so your thinking and eye for detail is good.

tippyweb replied:
Hello @nowucca 

Yes, I will check all pages again and update you later today.
Thank you so much for your patience!!

tippyweb replied:
Hello @nowucca 

One quick question:

This is from customers.jsp.

    _<c:forEach var="customer" items="${p.customerList}" varStatus="iter">

        <tr class="${((iter.index % 2) == 1) ? 'lightBlue' : 'white'} tableRow"
            onclick="document.location.href='customer/${customer.customerId}'">

                <%-- Below anchor tags are provided in case JavaScript is disabled --%>
            <td><a href="<c:url value="/admin/customer/${customer.customerId}"/>" class="noDecoration">${customer.customerId}</a></td>
            <td><a href="<c:url value="/admin/customer/${customer.customerId}"/>" class="noDecoration">${customer.name}</a></td>
            <td><a href="<c:url value="/admin/customer/${customer.customerId}"/>" class="noDecoration">${customer.email}</a></td>
            <td><a href="<c:url value="/admin/customer/${customer.customerId}"/>" class="noDecoration">${customer.phone}</a></td>
        </tr>

    </c:forEach>_

Inside the forEach loop, customer is a local variable, and so I can leave it as is.
Your confirmation is appreciated.

Thank you so much!!

tippyweb replied:
Hello @nowucca 

I looked into main pages where loop variables were handled in the same way.
So, I left those loop variables as is.

Updating the previous pull request shortly.  Thank you!!

nowucca commented on src/main/java/viewmodel/admin/BaseAdminViewModel.java:
```diff
@@ -0,0 +1,87 @@
+public class BaseAdminViewModel {
+
+    // The relative path to product images
+    private static final String PRODUCT_IMAGE_PATH = "/img/products/";
```
Can we share this with BaseViewModel.  ie  PRODUCT_IMAGE_PATH=BaseViewModel.PRODUCT_IMAGE_PATH ?

tippyweb replied:
Hello @nowucca 

Thank you so much for your quick review!!
Yes, it sounds great!!  That makes more sense.
Thank you for your suggestion, @nowucca !!!

tippyweb replied:
Good morning, @nowucca 

I tried BaseViewModel.PRODUCT_IMAGE_PATH by creating a new GET method in BaseViewModel as below:
    public String getPRODUCT_IMAGE_PATH () {
        return PRODUCT_IMAGE_PATH;
    }

But BaseViewModel.PRODUCT_IMAGE_PATH in BaseAdminViewModel doesn't get resolved.  Or we can make PRODUCT_IMAGE_PATH "protected", not "private", in BaseViewModel, which worked fine.  Would this be a better solution??  Not sure...  Thank you!!

nowucca replied:
Making it PUBLIC would work, or protected.  Basically I just wanted the constants chained together directly without any more getters etc.

tippyweb replied:
Great!  I used protected.  Thank you @nowucca !!

tippyweb replied:
@nowucca 

I attempted to do the same way for the delivery surcharge, but since it's not a static value, I couldn't do it.  Please take a look at my updates.  Thank you again and have a great weekend!!!

nowucca reviewed (approved):
tippyweb, this looks so great as it is.
I'm going to go ahead and test it, and merge it.

tippyweb committed:
Admin View Model for #29 Phase 2 updates

I found one initParam inside cart.jsp, which was updated with "p." too.
Thank you for your review in advance!!

nowucca commented:
Thank you so much @tippyweb for these changes.

tippyweb commented:
Good morning @nowucca 

This is wonderful.  Thank you so much again for your kind assistance, @nowucca !!
From this weekend, I'm taking SANS course "Defending Web Applications Security Essentials".  I hope to be able to do more volunteer work for you real soon again!!
Have a great week, @nowucca !!!
